R/chat_app.R

```diff
@@ -0,0 +1,216 @@
+#'   `messages` in [chat_ui()].
+#' @export
+chat_mod_ui <- function(id, ..., client = NULL, messages = NULL) {
+  if (!is.null(client)) {
+    check_ellmer_chat(client)
+
+    client_msgs <- map(client$get_turns(), function(turn) {
+      content <- ellmer::contents_markdown(turn)
+      if (is.null(content) || identical(content, "")) {
+        return(NULL)
+      }
+      list(role = turn@role, content = content)
+    })
+    client_msgs <- compact(client_msgs)
+
+    if (length(client_msgs)) {
+      if (!is.null(messages)) {
+        warn(
+          "`client` was provided and has initial messages, `messages` will be ignored."
+        )
+      }
+      messages <- client_msgs
+    }
+  }
+
+  shinychat::chat_ui(
+    shiny::NS(id, "chat"),
+    messages = messages,
+    ...
+  )
+}
+
+#' @describeIn chat_app A simple chat app module server.
+#' @export
+chat_mod_server <- function(id, client) {
+  check_ellmer_chat(client)
+
+  append_stream_task <- shiny::ExtendedTask$new(
+    function(client, ui_id, user_input) {
+      promises::promise(function(resolve, reject) {
+        tryCatch(
+          resolve(client$stream_async(user_input)),
+          error = function(err) {
+            reject(err)
+          }
+        )
+      })$then(function(stream) {
+        chat_append(ui_id, stream)
+      })
+    }
+  )
+
+  shiny::moduleServer(id, function(input, output, session) {
+    shiny::observeEvent(input$chat_user_input, {
+      append_stream_task$invoke(
+        client,
+        session$ns("chat"),
+        input$chat_user_input
+      )
+    })
+
+    shiny::observe({
+      if (append_stream_task$status() == "error") {
+        tryCatch(
+          append_stream_task$result(),
+          error = notify_error(session$ns("chat"), session)
+        )
+      }
+    })
+
+    shiny::reactive({
+      if (append_stream_task$status() == "success") {
+        client$last_turn()
+      }
+    })
+  })
+}
+
+notify_error <- function(id, session = shiny::getDefaultReactiveDomain()) {
+  function(err) {
+    needs_sanitized <-
+      isTRUE(getOption("shiny.sanitize.errors")) &&
+      !inherits(err, "shiny.custom.error")
+    if (needs_sanitized) {
+      msg <- "**An error occurred.** Please try again or contact the app author."
+    } else {
+      msg <- sprintf(
+        "**An error occurred:**\n\n```\n%s\n```",
+        strip_ansi(conditionMessage(err))
+      )
+    }
+
+    chat_append_message(
+      id,
+      msg = list(role = "assistant", content = msg),
+      chunk = TRUE,
+      operation = "append",
+      session = session
+    )
+    chat_append_message(
+      id,
+      list(role = "assistant", content = ""),
+      chunk = "end",
+      operation = "append",
+      session = session
+    )
+  }
+}
+
+strip_ansi <- function(text) {
+  # Matches codes like "\x1B[31;43m", "\x1B[1;3;4m"
+  ansi_pattern <- "(\x1B|\x033)\\[[0-9;?=<>]*[@-~]"
+  gsub(ansi_pattern, "", text)
+}
```

Comment:
Do you know, are these color codes?

Reply:
Yeah mostly colors or styles. They show up in the `cli` errors that are common now. They're hard to turn off with options (and we probably wouldn't want to so we get the formatting in the console), so it's just easier to strip them with regex.